### ISSUE TITLE
treelogic keys should be branches, not trees themselves

### ISSUE THREAD

**Issue by dwhswenson** [choderalab/openpathsampling#288]
Why are we giving entire subtrees as keys? I thought the keys were only supposed to be the actual path visited to get to a given mover.

Both approaches give unique keys, but the current version gives unreadable keys. Example: consider the following tree (discussed in a previous conversation):

![image](https://cloud.githubusercontent.com/assets/8178546/8312103/2054fbd6-19db-11e5-8623-1dc75f164ec2.png)

I would expect the blue path to give me

```
[Root, [RandomChoice, [ShootB]]]
```

Instead, for `Root.key(ShootB)`, I'm getting something like

```
[Root, [Sequential, [AxB, ShootA, ShootB, BxA]], [RandomChoice, [ShootA, ShootB]]]
```

This does NOT help me trace the path through the tree to get to that mover!

I mean, here's a real example:

```
>>> print root.key(self.movers['repex'][-1])
[<openpathsampling.pathmover.RandomChoiceMover object at 0x10bbe4490>, 
[<openpathsampling.pathmover.RandomChoiceMover object at 0x10bd8c0d0>,
[<openpathsampling.pathmover.PathReversalMover object at 0x10bbe4c90>], 
[<openpathsampling.pathmover.PathReversalMover object at 0x10bbe4fd0>], 
[<openpathsampling.pathmover.PathReversalMover object at 0x10bbe4d10>],
[<openpathsampling.pathmover.PathReversalMover object at 0x10bbe4d50>], 
[<openpathsampling.pathmover.PathReversalMover object at 0x10bbe4e10>], 
[<openpathsampling.pathmover.PathReversalMover object at 0x10bbe4cd0>], 
[<openpathsampling.pathmover.PathReversalMover object at 0x10bd8c090>]], 
[<openpathsampling.pathmover.RandomChoiceMover object at 0x10bd8c150>, 
[<openpathsampling.pathmover.OneWayShootingMover object at 0x10bbe3f50>, 
[<openpathsampling.pathmover.ForwardShootMover object at 0x10bbe3f90>], 
[<openpathsampling.pathmover.BackwardShootMover object at 0x10bd8c050>]]], 
[<openpathsampling.pathmover.RandomChoiceMover object at 0x10bd8c190>, 
[<openpathsampling.pathmover.OneWayShootingMover object at 0x10bbe4ad0>, 
[<openpathsampling.pathmover.ForwardShootMover object at 0x10bbe4a50>], 
[<openpathsampling.pathmover.BackwardShootMover object at 0x10bbe4f90>]], 
[<openpathsampling.pathmover.OneWayShootingMover object at 0x10bbe4c50>, 
[<openpathsampling.pathmover.ForwardShootMover object at 0x10bbe4910>], 
[<openpathsampling.pathmover.BackwardShootMover object at 0x10bbe4c10>]], 
[<openpathsampling.pathmover.OneWayShootingMover object at 0x10bbe4bd0>, 
[<openpathsampling.pathmover.ForwardShootMover object at 0x10bbe4990>], 
[<openpathsampling.pathmover.BackwardShootMover object at 0x10bbe48d0>]], 
[<openpathsampling.pathmover.OneWayShootingMover object at 0x10bbe4890>, 
[<openpathsampling.pathmover.ForwardShootMover object at 0x10bbe4a10>], 
[<openpathsampling.pathmover.BackwardShootMover object at 0x10bbe4a90>]], 
[<openpathsampling.pathmover.OneWayShootingMover object at 0x10bbe49d0>, 
[<openpathsampling.pathmover.ForwardShootMover object at 0x10bbe4950>], 
[<openpathsampling.pathmover.BackwardShootMover object at 0x10bbe4b10>]], 
[<openpathsampling.pathmover.OneWayShootingMover object at 0x10bbe4f50>, 
[<openpathsampling.pathmover.ForwardShootMover object at 0x10bbe4390>], 
[<openpathsampling.pathmover.BackwardShootMover object at 0x10bbe4ed0>]]], 
[<openpathsampling.pathmover.RandomChoiceMover object at 0x10bd8c1d0>, 
[<openpathsampling.pathmover.MinusMover object at 0x10bbe4d90>, 
[<openpathsampling.pathmover.EnsembleFilterMover object at 0x10bbd6050>, 
[<openpathsampling.pathmover.ConditionalSequentialMover object at 0x10bbd6310>, 
[<openpathsampling.pathmover.RandomChoiceMover object at 0x10bbe4dd0>, 
[<openpathsampling.pathmover.FirstSubtrajectorySelectMover object at 0x10bbe4e50>], 
[<openpathsampling.pathmover.FinalSubtrajectorySelectMover object at 0x10bbe4f10>]], 
[<openpathsampling.pathmover.RandomChoiceMover object at 0x10bbe4e90>, 
[<openpathsampling.pathmover.ReplicaExchangeMover object at 0x10bbe4450>]], 
[<openpathsampling.pathmover.RandomChoiceMover object at 0x10bbd6350>, 
[<openpathsampling.pathmover.ForwardExtendMover object at 0x10bbd6290>], 
[<openpathsampling.pathmover.BackwardExtendMover object at 0x10bbd62d0>]]]]], 
[<openpathsampling.pathmover.MinusMover object at 0x10bbd6550>, 
[<openpathsampling.pathmover.EnsembleFilterMover object at 0x10bbe3fd0>, 
[<openpathsampling.pathmover.ConditionalSequentialMover object at 0x10bbe33d0>, 
[<openpathsampling.pathmover.RandomChoiceMover object at 0x10bbd6150>, 
[<openpathsampling.pathmover.FirstSubtrajectorySelectMover object at 0x10bbd60d0>], 
[<openpathsampling.pathmover.FinalSubtrajectorySelectMover object at 0x10bbd6110>]], 
[<openpathsampling.pathmover.RandomChoiceMover object at 0x10bbd6190>, 
[<openpathsampling.pathmover.ReplicaExchangeMover object at 0x10bbd61d0>]], 
[<openpathsampling.pathmover.RandomChoiceMover object at 0x10bbe3450>, 
[<openpathsampling.pathmover.ForwardExtendMover object at 0x10bbd6210>], 
[<openpathsampling.pathmover.BackwardExtendMover object at 0x10bbd6250>]]]]]], 
[<openpathsampling.pathmover.RandomChoiceMover object at 0x10bd8c210>, 
[<openpathsampling.pathmover.ReplicaExchangeMover object at 0x10bbe4690>], 
[<openpathsampling.pathmover.ReplicaExchangeMover object at 0x10bbe45d0>], 
[<openpathsampling.pathmover.ReplicaExchangeMover object at 0x10bbe4550>], 
[<openpathsampling.pathmover.ReplicaExchangeMover object at 0x10bbe4710>], 
[<openpathsampling.pathmover.ReplicaExchangeMover object at 0x10bbe4610>], 
[<openpathsampling.pathmover.ReplicaExchangeMover object at 0x10bbe4b50>]]]
```

I would have expected:

```
[<openpathsampling.pathmover.RandomChoiceMover object at 0x10bbe4490>,
[<openpathsampling.pathmover.RandomChoiceMover object at 0x10bd8c210>, 
[<openpathsampling.pathmover.ReplicaExchangeMover object at 0x10bbe4b50>]]]
```

which is a lot more useful.

**Comment by dwhswenson** [comment]
I should clarify that, knowing the path, we get a unique key. A given mover may appear more than once in the tree, in which case the mover does not HAVE a unique key in either case. However, identifying all the keys for any given mover is as simple as:

1. Enumerate all total branches. This scales as the number of leaf nodes; not ideal, but within reach for anything we'd be doing.
2. Keep all branches which contain the desired mover.
3. Truncate remaining branches at the desired mover.
4. Remove duplicates.

**Comment by jhprinz** [comment]
You are right. This is not optimal :) (potentially even very far from being optimal).
The current implementation was simple and creates a 'subtree' that will work as a unique key under all circumstances. This was the primary goal.

It is definitely a good idea to improve on that.

[the following part needs to be worked on a little. I just write some stuff down so I don't forget about it]

I remember vaguely that we had a reason to make it that complicated, but it could be that these were some very pathological cases that we do not need anymore.

> This does NOT help me trace the path through the tree to get to that mover!

Do you mean it is too long? I agree, but I will get you to the right node.

Let me recapitulate:

1. We want to have a unique key for each mover in a movertree
2. Check whether  any of the mover from the movetree appears in a particular path
3. It is desirable, if we can also 'guess' the key as you did it in your example above.

This works with the current scheme, but results in very long unique keys.

What should work right now:
1. You can get the unique key using `.key(node)`, but this will only work if a node appears only once. Otherwise you cannot guess from the mover alone which particular node in the tree is meant
2. You can check, if a subtree is in the path, this could help make it easier and potentially avoid having to use the unique key (which might not even be unique)

What I have in mind is this. Imagine that your analysis object has a reference to a MinusMover `minus`, which might be used several times (for whatever reason). Then, what good does a unique key do for the MinusMover? The three questions that seem to make sense are:
1. Was the MinusMover used in a particular pathmovechange `path` as all, which would be checked using `minus in path`
2. Was a particular Mover used in `path`, which we want to check using the unique key. The problem is, that there is no direct way to pick a particular unique key. At least not by selecting a mover like `minus`, because it does not have a unique key. So if a user really wants to pick a particular one, how would he get the unique key.

    ´´´
    mover = Seqential([shooterA, shooterA])
    ´´´
    And we want to know how often the second mover is accepted. How to get the unique key? You should be able to guess just like `[Sequential, [shooterA], [shooterA]]`
3. You are interested in the statistics among all possible calls of `minus`, in which case it makes sense to have an automatic listing of all unique keys that involve a particular mover like `minus`. E.g. `root.keys(minus)`.

Finally, I think at some point we agreed that we should be able to select a node by using a subtree and if the subtree matches we pick the first place where the subtree fits in the full tree and the last node visited from the subtree is the node we were looking for. The 4 steps you describe above should work to get a minimal working subtree for each node.

**Comment by dwhswenson** [comment]
Two points here (one short, one long):

#### What got me asking about this

What brought this up for me: I was trying to find a way to get the probability of choosing a certain move by tracing until I ran out of RandomChoice movers. This also serves as a good definition of a "canonical" move. I *thought* we'd agreed on keys that would make that easy (the sort I gave as examples above).

In the end, I abandoned the goal of a generic calculation of the move choice weight and instead require that all `GLOBAL` move strategies assign it themselves, and then try to preserve it when they set themselves up. So that purpose is no longer needed.

#### Remaining concerns

Here are my concerns. The primary purpose to these keys are to unambiguously identify a path (branch) through the move decision tree. Yes, what you have does that. So would a cryptographic hash of the same thing. The problem with the current implementation is that it looks like it *should* have meaning in it, but the meaning is easily lost.

For example, look at the replica exchange move I gave as a real example. If I ask the question, "in this path, did I do a shooting move at some point?" my intuitive response, looking at that key, would be "yes -- look at all the shooting moves that appear in this key!" That, in a nutshell, it what is wrong with this implementation.

Another point is that if the order of movers in a random choice mover changed, the keys would change. I wouldn't expect that if the key had real meaning (beyond serving the role of a hash). 

You seem to say that your three "recapitulation" points work in the current scheme. I don't see how (2) or (3) work. One or more of three things is happening here:

1. We're using "path" differently. I mean the nodes visited by the blue arrow in my picture above. That's what I'm calling a "path" (or "branch") in this context. I do not think I can easily get that out of the current scheme.
2. You mean "implementable" instead of "implemented" when you say "works with the current scheme." That may be.
3. There's some part of the tree mix-in that I haven't found yet.


I'm a bit uncertain about the entirety of including the tree mix-in with PathMovers. I know there was some part of it you wanted, but there are a lot of other parts that don't make sense. For example, `__iter__` traverses the tree. This is a dangerous kind of nonsense, because it can lead to all sorts of errors. Already, I've had a half-dozen times that iterating over `x in mover` (when I meant `x in mover.movers` or where I'd thought `mover` was a list of movers) caused some very strange behavior (all the shooting movers are in a replica exchange mover???) -- this should have raised an error, because the resulting behavior is not something anyone would expect or desire from the native iteration.

**Comment by jhprinz** [comment]
Hmm, what you describe is not what we agreed upon to implement. 
Also, in the code the key is not used right now, is that correct?

> Here are my concerns. The primary purpose to these keys are to unambiguously identify a path (branch) through the move decision tree.

No, that is not how I recall it and that might be the reason of the problems. As far as I remember a key if a unique way to identify a particular mover not a path to it to the root. Problem is that the path to a mover can depend of all that happens before. That is the reason for the huge key.

I really think we are using path in a different way. Okay, let me give this another try. Let's use your example question

> For example, look at the replica exchange move I gave as a real example. If I ask the question, "in this path, did I do a shooting move at some point?" my intuitive response, looking at that key, would be "yes -- look at all the shooting moves that appear in this key!" That, in a nutshell, it what is wrong with this implementation.

I would respond to write
```
if shooting in path:
    # do something
```
should solve your problem. So, let's ask, what is the ratio of left shootA vs right shootA.

I would write
```
if [Root, [RandomChoice, [ShootB]]] in path:
    # right side

if [Root, [Sequential, [AxB], [ShootB]]] in path:
    # left side
```

Now, I understand that you are looking for these keys that I used, correct? And you want to be able to get a key or list of keys for all of these. Problem is that these are ill defined because the occurance of a mover depends on the past. So there can be several ways to get to a specific mover. For each of these you can specify the complete past.

Problem that I see right now is that we cannot (yet) evaluate all possible pasts for a specific mover which is simple but not implemented. This would have to happen at PathMover level and
`root.keys(shootA)` would return two subtrees/paths that lead unambigeously to that mover. This will involve the complete past. It can be done using tha algorithm you wrote above.

If that is the problem I will create a PR to fix it.

**Comment by dwhswenson** [comment]
> I really think we are using path in a different way.

Can you please explain what your definition of "path" is?

In the example above, for me, there are two possible "paths" including `ShootB`

```
[Root, [Sequential, [AxB, ShootA, ShootB, BxA]]]
```
and
```
[Root, [RandomChoice, [ShootB]]]
```

I do not understand your usage of `in` above. Note that, in the keys we have (which, for things on the right side, list everything on the left as well), the key for the right side would be something like `[Root, [Sequential, [AxB, ShootA, ShootB, BxA]], [RandomChoice, [ShootA, ShootB]]]`. Isn't `[Root, [Sequential, [AxB, ShootA, ShootB]]]` "`in`" that? I would think that `[Root, [Sequential, [AxB, ShootA, ShootB]]]` would be "`in`" `[Root, [Sequential, [AxB, ShootA, ShootB, BxA]]]`: if any subtree is "`in`" a global tree, then you can't use "`in`" to distinguish two branches including the same mover unless you know which is left and which is right. And since left and right are only internal constructs, I think that's a bad way of implementing it.

> ```
> if shooting in path:
>     # do something
> ```

Again, where is this `path`? Is this the key? Is `<openpathsampling.pathmover.OneWayShootingMover object at 0x10bbe3f50>` not "`in`" the key I show above? It definitely shows up in both the printing of the key and in the `__iter__` of the key; not sure about `__contains__`, but that should be consistent. This suggests to me that the shooting mover is "in" the replica exchange mover. This is exactly what I'm arguing is nonsense.

**Comment by dwhswenson** [comment]
Just to clarify why `__contains__` needs to be consistent with `__iter__` (which I'm not sure is the case with the tree logic stuff):

```python
for element in iterable:
    assert(element in iterable)
```

If this doesn't pass, then we've made something terribly confusing.

**Comment by dwhswenson** [comment]
(Sorry, one more note:)

> Problem that I see right now is that we cannot (yet) evaluate all possible pasts for a specific mover which is simple but not implemented. This would have to happen at PathMover level and
root.keys(shootA) would return two subtrees/paths that lead unambigeously to that mover. This will involve the complete past. It can be done using tha algorithm you wrote above.

Not actually that simple. My suggestion is to, by default, ignore any RandomChoiceMovers within a SequentialMover. My claims of it being computationally easy are true until you have long sequences of random choices: over that range, the number of paths scales as $N^S$ where $S$ is the number of steps in the sequence and $N$ is the number of choices in the RandomChoice (with the obvious generalization if $N$ isn't constant). Of course, that number can be determined, and you can raise an error if the user tries to enumerate a ridiculously large number of such paths.

**Comment by jhprinz** [comment]
Yes, I just ran into that problem, that if you have lots of RandomChoiceMovers there will be lots of possibilities :) 

> Just to clarify why __contains__ needs to be consistent with __iter__ (which I'm not sure is the case with the tree logic stuff):

Yes, you are right, that should be the case and the way this works right now is not the case. `__iter__` iterates all nodes, but contains (so far) only checks for subtrees starting from the root. So your tests fails!

Let me try to come up with a solution to this key/tree, etc mess...
